Display chamber temperature

GMagician opened the pull request:
This fix some compile errors introduced by #14190, and also give the possibility to display chamber temperature when chamber is not heated.

Still to decide icon X position

GMagician committed:
Display chamber temperature

This fix some compile errors introduced by #14190, and also give the possibility to display chamber temperature when chamber is not heated.
Still to decide icon X position

thinkyhead committed:
Cleanups to updated status screen

GMagician commented on Marlin/src/lcd/dogm/status_screen_DOGM.cpp:
```diff
@@ -72,7 +72,7 @@ FORCE_INLINE void _draw_centered_temp(const int16_t temp, const uint8_t tx, cons
 #define DO_DRAW_BED (HAS_HEATED_BED && STATUS_BED_WIDTH && HOTENDS <= 3 && DISABLED(STATUS_COMBINE_HEATERS))
 #define DO_DRAW_FAN (HAS_FAN0 && STATUS_FAN_WIDTH && STATUS_FAN_FRAMES)
-//#define DO_DRAW_CHAMBER (HAS_TEMP_CHAMBER && ((HOTENDS <= 2 && DO_DRAW_BED) || (!DO_DRAW_BED && HOTENDS <= 3)))
+#define DO_DRAW_CHAMBER (HAS_TEMP_CHAMBER && ((HOTENDS <= 2 && DO_DRAW_BED) || (!DO_DRAW_BED && HOTENDS <= 3)))
```
This was disabled because currently chamber icon is displayed over the bed icon...
I think that X may be left of bed (if possible) or over bed location when this is not present

GMagician replied:
Fixing all this

GMagician commented on Marlin/src/lcd/dogm/status_screen_DOGM.cpp:
```diff
@@ -362,7 +362,7 @@ void MarlinUI::draw_status_screen() {
 
   #if DO_DRAW_CHAMBER
-    #if HAS_HEATED_CHAMBER && defined(STATUS_CHAMBER_ANIM)
+    #if HAS_HEATED_CHAMBER && ENABLED(STATUS_CHAMBER_ANIM)
```
Following your `STATUS_CHAMBER_ANIM` undef when no heated chamber, this may be `#if ENABLED(STATUS_CHAMBER_ANIM)`, I think

GMagician commented on Marlin/src/lcd/dogm/status_screen_DOGM.cpp:
```diff
@@ -72,7 +72,7 @@ FORCE_INLINE void _draw_centered_temp(const int16_t temp, const uint8_t tx, cons
 #define DO_DRAW_BED (HAS_HEATED_BED && STATUS_BED_WIDTH && HOTENDS <= 3 && DISABLED(STATUS_COMBINE_HEATERS))
 #define DO_DRAW_FAN (HAS_FAN0 && STATUS_FAN_WIDTH && STATUS_FAN_FRAMES)
-//#define DO_DRAW_CHAMBER (HAS_TEMP_CHAMBER && ((HOTENDS <= 2 && DO_DRAW_BED) || (!DO_DRAW_BED && HOTENDS <= 3)))
+#define DO_DRAW_CHAMBER (HAS_TEMP_CHAMBER && ((HOTENDS <= 2 && DO_DRAW_BED) || (!DO_DRAW_BED && HOTENDS <= 3)))
```
I just saw now a define that may help:
#define DO_DRAW_CHAMBER (HAS_TEMP_CHAMBER && (HOTENDS == 1 || HOTENDS < MAX_HOTEND_BITMAPS))

thinkyhead replied:
Follow the bed model as closely as possible. Note especially that the chamber should not be drawn if `ENABLED(STATUS_COMBINE_HEATERS)`, because in that case the chamber will be included in the single header graphic already.

GMagician replied:
Doing that in following commit

GMagician reviewed (commented):
I wrote some comments on your changes

GMagician committed:
Add chamber X pos computation

GMagician commented:
Now icons X is computed correctly. Only "bad" thing is that this code consider right but not left padding.
Solution 1: left align all icons so no left padding is present (cons: rightmost icon may leave too much empty space on screen right)
Solution 2: add a LEFT_PADDING_WIDTH to each iconset ad use in `ICON_SPACING` macro

GMagician committed:
Fix lost change

thinkyhead committed:
Fix up DOGM status screen

thinkyhead committed:
Group status screen conditionals

thinkyhead committed:
Try separate hotend/bed drawing

GMagician commented:
@thinkyhead what do you think is best solution for icon spacing?
Left align images or add a LEFT_PADDING define and handle it in `ICON_SPACING` macro?
Now chamber icon is 16 pixels wide and fan has 2 padding pixels on left so no extra-space should be added. This code now puts 8+2 pixels between those icons...not a good results

thinkyhead commented:
> not a good results

Indeed. It is optimal for a bitmap to be on a byte boundary, but it isn't absolutely required. When it really matters to performance, the single header bitmap and simple heating indicators should be used.

A primary consideration is that each indicator needs enough horizontal space for the temperature reading and some whitespace — around 24 pixels width. The chamber probably will not go to 100° and up, but we might still want to reserve space for it, in which case the chamber icon should be modified to take up as much width as the bed with the same extra padding.

The positioning logic doesn't have to be too complicated, but it could do a basic heuristic: If the icons can all fit when simply positioned 24 pixels apart, keeping to byte boundaries, that should be used. If they cannot, maybe fall back to using the defined icon width for each one, plus two pixels of space (still keeping about 24 pixels width minimum for the temperature), or figure out how to center-align them to the temperature text (using `STATUS_HOTEND_TEXT_X`, etc., to infer the left padding in the bitmap).

The fan should basically always be drawn at 104 (`LCD_PIXEL_WIDTH - 24`). It would also be fine to assume the first nozzle should be drawn at the byte boundary after the logo bitmap (if any).

It would also makes sense to not draw the status logo image when it would prevent the second extruder, bed, or chamber from being drawn.

Simplest approach is just to divide up the space from the right edge of the logo to the left edge of the fan, divide by 24, and figure out to do with those slots. If they aren't enough, take zero as the left edge and make use of the extra slot or two.

GMagician commented:
Please I may have problem to translate correctly your statement, please forgive me if I write weird things.
Analyzing what you say every icon should display up to 3 digits temps (temp is 'nnn°' enough 24 pixels?) and at least 2 pixel more for readability (left padding).
Icons bmp, to save memory, are not always "2 empty pixels left padded" (alt bed and chamber are an example) so image X should step by variable values:
  26 for fan should be ok since real image will starts at 104: 128-26=102 + 2 padding in bmp
  26 for alt bed is not ok since image will start at 78 and will be printed at 78 (no padding in bmp) and text centring logic has to know these 2 pixels are missing (that's why you defined a TEXT_X for bed), otherwise text will not be centred
Now I think about bmp > 24 wide, things are more complicated. Stepping should be MAX(26, iconw) and again if padding spaces are present in bmp knowing that is more important to realize a correct stepping.
Finally this variable stepping will complicate logic to know if logo has to be printed or not.For each object computation is needed to know its slot reserved pixels. Adding "hotend" slots size we know if bed/chamber may be displayed (now is simply used HOTEND number) and even if logo has to be sacrificed...

GMagician committed:
Slot aligned status icons

-Each status icon use now a 24 pixel wide slot.
  Images and its text are centered in these slots
  Bad may have an exception due to PERCENT BAR
- Used a single define for all "target" text Y
- Changed temperature draw centered text, now '°' char is also considered
- All icons are now left aligned (no more left padding is needed)
- Saved some progmem when no STATUS_HEAT_PERCENT is not enabled

Maybe in future chamber will need also a cooling option. With bed heated for a lot of hours temperature may also need to be lowered and not heated....

thinkyhead commented:
> 26 for fan should be ok since real image will starts at 104: 128-26=102 + 2 padding in bmp

Fan graphics should be right-aligned so that the fan can always be 24 pixels width with empty space on the left. Thus, the other graphics can extend all the way from pixel 0 to pixel 103 when there is a fan. When there is no fan, they can extend all the way to 127.

~~On the left... If there is no LOGO image and the first heater was drawn at pixel 0, 3 digits of text should be able to fit below it and appear centered, with a little extra space on the right for the `°` character.~~

We currently have macros like `STATUS_HOTEND_WIDTH`, `STATUS_HOTEND_X`, etc., which users can define for situations where they want to get nicer spacing for their custom graphics. This was all done so we don't have to add a bunch of logic. So we might not need to do anything.

thinkyhead commented:
Yeah… Let's remove `ICON_SLOT_WIDTH` and `ICON_SLOT_SPACING` and forget the idea about having a general adaptable layout for now. Graphics can vary a lot, and we don't expect users to include a lot of meta-data. A width and a text offset position for each icon is what we now use, and this works well to control positioning and allow for user control. (i.e., the less automation the better here)

The status screen layout is admittedly very basic. It is meant to be more of an example that's relatively easy to follow and hack. We don't need to extend it into a full-fledged system.

The status screen files themselves decide what to display based on number of extruders, and so on, and can hide the logo at their choosing. So we don't need anything new here.

I would like it if users would bring us a variety of custom graphical status screen layout ideas. And the ones that are particularly good, we should add to Marlin as options. There is a lot of room for experimentation and improvement!

GMagician commented:
> Let's remove ICON_SLOT_WIDTH and ICON_SLOT_SPACING and forget the idea about having a general adaptable layout for now

The idea of these defines is not only for an "adaptable" layout but also because of code maintainability. But I can expand their content into code if you like it...let me know

GMagician commented:
[OT] Sorry @thinkyhead I don't know how to reach developpers (I asked something, not this, on discord but no answer yet), this is just a programmer C/C++ question...
I'm playing with marlin for a new board HAL (don't know where I will go with it) I'm wondering if there is a possibility to initialize a static variable with a pointer in compile time...
eg
char myarray[10] = { 1, 2,  ....};
uint32_t mypointer = &myarray[5];
The real code is trying to initialize this into an array of struct where one field is a uint32_t but should store an address to something (this question is related to some DMA register initialization)
I think that this should be possible even if correct address will be known at link time.
Thanks for you time

thinkyhead commented:
> initialize a static variable with a pointer in compile time

Yes, that is fine. Static variables are simply in global data space with a single instance and can contain anything that a variable not marked `static` can contain.

GMagician commented:
> Yes, that is fine. Static variables are simply in global data space with a single instance and can contain anything that a variable not marked static can contain.

Not so fine for compiler... I get "Invalid initial value for member 'DSTADDR'"

Please note that DmacDescriptor is a struct and each field is a union
```c++
  #define ADC0_AINCOUNT   COUNT(adc1_dma_regs_list)

  uint16_t HAL_adc_results[COUNT(adc_pins)];

  DmacDescriptor dma_1st_descriptor[DMA_CHANNELS] = {
    #if ADC0_IS_REQUIRED
      // DMA_ADC0_RD_CHANNEL
      { { true,                                         // BTCTRL.VALID
          DMAC_BTCTRL_EVOSEL_DISABLE_Val,               // BTCTRL.EVOSEL
          DMAC_BTCTRL_BLOCKACT_NOACT_Val,               // BTCTRL.BLOCKACT
          DMAC_BTCTRL_BEATSIZE_HWORD_Val,               // BTCTRL.BEATSIZE
          false,                                        // BTCTRL.SRCINC
          true,                                         // BTCTRL.DSTINC
          false,                                        // BTCTRL.STEPSEL
          DMAC_BTCTRL_STEPSIZE_X1_Val },                // BTCTRL.STEPSIZE
        { ADC0_AINCOUNT },                              // BTCNT
        { (uint32_t)&ADC0->RESULT.reg },                // SRCADDR
        { (uint32_t)&HAL_adc_results[ADC0_AINCOUNT] },  // DSTADDR
        { 0 }                                           // DESCADDR
      },
  [cut]
```

You may find whole code here https://github.com/GMagician/Marlin/tree/grand-central-M4

thinkyhead commented:
> The idea of these defines is not only for an "adaptable" layout but also because of code maintainability. But I can expand their content into code if you like it...let me know

What I'm saying is that we should not put too much effort in the current direction. Status screens are meant to be adjustable so that users and vendors can get exactly the positions they want for the configurations they want to support, and nothing more. The idea that there is an "adaptable" layout here is kind of an illusion. The bitmaps are not "icons" and should not be implied to want or need uniform sizing or spacing. The layout does not "adapt" but simply provides reasonable defaults.

To give some historical perspective, the status screen header has always been a static layout, and I wasn't aiming to change that. The only thing I wanted to achieve with the recent rewrite was less duplication of effort and data, so that vendor status screens will only need to include a logo bitmap and all the other bitmaps come from Marlin. So when we update Marlin's graphics the vendor screens will inherit the new graphics too.

In the process, I just wanted a way for them to be able to set the positions of those bitmaps, so I settled on regular spacing for nozzles (as an option), and more explicit positions for bed / chamber / etc. And with the fan on the far right by default. I wanted all the positioning to have some sensible default fallbacks so as you develop your status screen, if you forget to define the position of the bed then it would still be visible and you could adjust from there. If the default happens to work, then you can leave it out of your custom status screen.

I would like to avoid trying to make any more of a "system" out of it, and instead focus on the graphical display code to make it cleaner and better. It is already evolving and improving and -time allowing- I want to do more experimentation with alternative screen layouts that could be much better. With more effort I know we could build much better status screens for Marlin 2.1 and I would like to see the current status screen code go away completely. Knowing as I do how "bad" the current code is, I feel like we are simply rearranging the deck chairs on the Titanic.

So let us simply focus on adding the Heated Chamber to the screen under the current system and let's not try to change it here. This PR should have been done and merged long ago, but it has gotten bogged down by concepts. Nothing poisons and delays a PR worse than concepts.

GMagician commented:
Ok thinkyhead as per your wish :-D
I didn't meant to change current logic (you can still force bitmap positions by defining proper defines), just some automation to avoid formulas replication, I'll revert to that...

GMagician commented:
I tried to do something different but result is ugly...too much unused space, bitmap too distant...I prefer to leave old implementation (without chamber icon)

thinkyhead commented:
We could probably do a nice layout with temps side-by-side so we can fit more data. Something like…
```
1: 67/100° B: 21/  0° F: 98%
2: 22/  0° C: 44/ 45°
```
…where the labels are replaced by smaller (8x8) icons. And the icons can have multiple images, with one for OFF, one for HEATING, etc. Unfortunately a horizontal temperature indicator takes up about 9 characters, so you could only fit two across.

The XYZ indicator in the center of the screen may not be in the best position, and as many have pointed out, the XY indicator is not useful information during a print job. If the XY display was removed entirely, that would simplify things, but some like to have it around when paused or when outside of a print job.

It has been suggested before that during printing the XY should be hidden and the display should be  narrowed to the status information relevant to the print job, even adding on the total filament usage, for example. I would love to see some design suggestions that take advantage of the extra space freed up by removing XY, and which make the progress of the job easier to see at a glance.

Magnified numbers the size of the screen that cycle…? Circular progress pie instead of progress bar…? What more can we do to pretty up the status screen?

GMagician commented:
Please note that font is, on my display at least, 6x? pixels, so 8x8 is bigger!

The proposed code was working optimizing icon distance and leave coders to put icons in a preferred position (forcing x pos). It also skip icons if space is not enough. It was less dramatic than solution you are proposing here (this seems like the 2.1 revolution you was writing about).
The problem I faced, rolling back to 'no slot_width', was the fact that code has fixed X computation and since icons may or may not be present will lead to complicate define game. Considering that icons may also be multiple of 8 pixels and that bed may have percent bar, it will complicate things again.
The PR was the best solution I found to solve all these problems. I didn't found another valid approach, but I'm ready to give it another try

GMagician commented:
>  I would love to see some design suggestions that take advantage of the ...

I think I'm not a good "designer" to propose something about that

boelle commented:
if you need an idea for an icon take a icon of a printer and put it inside a box and take the heat waves from the heated bed icon and put that under it or inside it

GMagician commented:
@boelle the icon is still present, for heated chamber, problem here arise to positioning logic

boelle commented:
so we have an icon for heated chamber like for heated bed and hotend?

but i would place it last, ie next to heated bed

or are we talking about text displays?

GMagician commented:
> so we have an icon for heated chamber like for heated bed and hotend?

yes

> but i would place it last, ie next to heated bed

this pr draw, from right, fan, bed, chamber (no prefered slot, if some icon is missing others are shifted) and if too many extruders, chamber may be hidden

> or are we talking about text displays?

@thinkyhead had some complain about the changes wrote on this PR, too much "automation" on icon positioning. The real issues is I'm not able to find a different, simpler, solution. @thinkyhead proposed a new layout changing more things (text and screen content) and wrote:

>  I would love to see some design suggestions that take advantage of the extra space freed up by removing XY, and which make the progress of the job easier to see at a glance.

that followed my "I'm not a good designer" statement

boelle commented:
yeah, at least i see the problem if user has a board than can have say 5 extruders and then bed, no room for the heated chamber

i think in that case it should place chamber last and kind of scroll the extruders so you see 2-3 extruders for 3 sec and then the others for 3 sec...

during print we could skip displaying the X Y cords as they change to much to be usefull

GMagician commented:
> yeah, at least i see the problem if user has a board than can have say 5 extruders and then bed, no room for the heated chamber

With current code also bed may be sacrificed when too many extruders are present, heated chamber act the same way

> i think in that case it should place chamber last and kind of scroll the extruders so you see 2-3 extruders for 3 sec and then the others for 3 sec...

that was far from what this PR would add...I meant just to display chamber temperature when possible

boelle commented:
hmm, could we maybe give user an config option on what icons they want all the time and which can be scrolled/changed and how often they should change them?

but yes, this might be better if split in to a new FR ?

AnHardt commented:
> I would like it if users would bring us a variety of custom graphical status screen layout ideas. And the ones that are particularly good, we should add to Marlin as options. There is a lot of room for experimentation and improvement!

Maybe some of the ideas from #5430 are still usable.

When out of display places consider (blinking) iterating thru a list of heaters (items) at one display place.
For example with 3 places:
* leftmost, fix, current extruder
* next from left, alternating, thru the extruders
* rightmost, alternating, part cooling fan, bed, chamber, environment temperature, humidity in filament storage box, weight of filament spool - tara spool, pie graphic of left printtime / total printtime, state of integrated coffee maker, high scores of games, days left to doom.

Fix the more important ones - what is a matter of taste.
Center in places. Group by similar width.
Or make it configurable in a array (list) of lists.
More important items could appear more than once in that lists.

Even the position and size of the display fields could be a configurable list of arrays - covering the whole display.
